File: week07/20/pr5_Search_a_2D_Matrix_2.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int idx = 0;
        for(int i=0;i<m;i++){
            if(matrix[i][0]>target) break;
            idx++;
        }
        if(idx==0) return false;
        for(int j=0;j<idx;j++){
            int left = 0;
            int right = n-1;
            while(left<=right){
                int mid = (left+right)/2;
                if(matrix[j][mid]==target) return true;
                else if(matrix[j][mid] < target) left = mid+1;
                else right = mid-1;
            }
        }
        return false;
    }
};

File: week07/20/pr5_Search_a_2D_Matrix_2.py
#😊
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        row_idx = 0
        for i in range(m):
            if matrix[i][0] >= target:
                if matrix[i][0] == target : return True
                break
            else : row_idx += 1
        if row_idx == -1 : return False
        
        for j in range(row_idx): # index check!
            # Binary Search for each row
            left = 0
            right = n-1
            while left <= right:
                mid = (left+right)//2
                if matrix[j][mid] == target : return True
                elif matrix[j][mid] > target : right = mid-1
                else : left = mid+1

        return False

